npm/pkg/dataplane/dataplane.go

```diff
@@ -366,52 +394,151 @@ func (dp *DataPlane) applyDataPlaneNow(context string) error {
 func (dp *DataPlane) AddPolicy(policy *policies.NPMNetworkPolicy) error {
 	klog.Infof("[DataPlane] Add Policy called for %s", policy.PolicyKey)
 
-	// Create and add references for Selector IPSets first
-	err := dp.createIPSetsAndReferences(policy.AllPodSelectorIPSets(), policy.PolicyKey, ipsets.SelectorType)
-	if err != nil {
-		klog.Infof("[DataPlane] error while adding Selector IPSet references: %s", err.Error())
-		return fmt.Errorf("[DataPlane] error while adding Selector IPSet references: %w", err)
+	if !dp.netPolInBackground {
+		return dp.addPolicies([]*policies.NPMNetworkPolicy{policy})
 	}
 
-	// Create and add references for Rule IPSets
-	err = dp.createIPSetsAndReferences(policy.RuleIPSets, policy.PolicyKey, ipsets.NetPolType)
-	if err != nil {
-		klog.Infof("[DataPlane] error while adding Rule IPSet references: %s", err.Error())
-		return fmt.Errorf("[DataPlane] error while adding Rule IPSet references: %w", err)
+	// Choose to keep netPolQueue locked while running iptables-restore within addPoliciesWithRetry.
+	// We are not blocking any thread but the background NetPol thread, which would run the same command anyways
+	dp.netPolQueue.Lock()
+	defer dp.netPolQueue.Unlock()
+
+	dp.netPolQueue.enqueue(policy)
+	newCount := dp.netPolQueue.len()
+
+	klog.Infof("[DataPlane] [%s] new pending netpol count: %d", contextAddNetPol, newCount)
+
+	if newCount >= dp.MaxPendingNetPols {
+		klog.Infof("[DataPlane] [%s] applying now since reached maximum batch count: %d", contextAddNetPol, newCount)
+		dp.addPoliciesWithRetry(contextAddNetPol)
 	}
+	return nil
+}
 
-	var endpointList map[string]string
-	if dp.inBootupPhase() {
-		// During bootup phase, the Pod controller will not be running.
-		// We don't need to worry about adding Policies to Endpoints, so we don't need IPSets in the kernel yet.
-		// Ideally, we get all NetworkPolicies in the cache before the Pod controller starts
-		err = dp.incrementBatchAndApplyIfNeeded(contextAddNetPol)
+// addPoliciesWithRetry tries adding all policies. If this fails, it tries adding policies one by one.
+// The caller must lock netPolQueue.
+func (dp *DataPlane) addPoliciesWithRetry(context string) {
+	netPols := dp.netPolQueue.dump()
+	klog.Infof("[DataPlane] adding policies %+v", netPols)
```

Comment:
Can we make this a debug statement ?

Reply:
Seems like we don't have debug logging capability via klog